include/Dialects/LinalgTransform/TransformOpInterface.h

```diff
@@ -0,0 +1,226 @@
+/// transformation to a single operation handle and producing a single operation
+/// handle. The op must implement:
+///   - FailureOr<convertible-to-Operation*> applyToOne(OpTy) - a transformation
+///   that is applied in turn to all payload IR operations that correspond to
+///   the handle of the transform IR operation.
+/// In the functions above, OpTy is either Operation * or a concrete payload IR
+/// Op class that the transformation is applied to (NOT the class of the
+/// transform IR op).
+template <typename OpTy>
+class TargetableOpTrait : public OpTrait::TraitBase<OpTy, TargetableOpTrait> {
+public:
+  /// Applies the transformation to each op from the only target and sets the
+  /// only result to correspond to the list of individual results.
+  LogicalResult apply(TransformResults &transformResults,
+                      TransformState &state) {
+    using TransformOpType = typename llvm::function_traits<
+        decltype(&OpTy::applyToOne)>::template arg_t<0>;
+    ArrayRef<Operation *> targets =
+        state.getTargets(this->getOperation()->getOperand(0));
+    SmallVector<Operation *> results;
+    if (failed(applyTransformToEach(
+            targets, results, [&](TransformOpType specificOp) {
+              return static_cast<OpTy *>(this)->applyToOne(specificOp);
+            })))
+      return failure();
+    transformResults.set(0, results);
+    return success();
+  }
+
+  /// Verifies that the op satisfies the requirements for this trait.
+  static LogicalResult verifyTrait(Operation *) {
+    static_assert(OpTy::template hasTrait<OpTrait::OneOperand>(),
+                  "expected single-operand op");
+    static_assert(OpTy::template hasTrait<OpTrait::OneResult>(),
+                  "expected single-result op");
+    return success();
+  }
+};
+
+/// Trait implementing the TransformOpInterface for operations applying a
+/// transformation to a single operation handle and not producing any handles.
+/// The op must implement:
+///   - LogicalResult applyToOne(OpTy) - a transformation that is applied in
+///   turn to all payload IR operations that correspond to the handle of the
+///   transform IR operation.
+/// In the functions above, OpTy is either Operation * or a concrete payload IR
+/// Op class that the transformation is applied to (NOT the class of the
+/// transform IR op).
+template <typename OpTy>
+class TargetableNoResultOpTrait
+    : public OpTrait::TraitBase<OpTy, TargetableNoResultOpTrait> {
+public:
+  /// Applies the transformation to each op from the only target.
+  LogicalResult apply(TransformResults &transformResults,
+                      TransformState &state) {
+    using TransformOpType = typename llvm::function_traits<
+        decltype(&OpTy::applyToOne)>::template arg_t<0>;
+    ArrayRef<Operation *> targets =
+        state.getTargets(this->getOperation()->getOperand(0));
+    SmallVector<Operation *> results;
+
+    for (Operation *target : targets) {
```

Comment:
Looking a bit above, it isn't clear to me that the duplication is needed.
Would it be too convoluted to have only 1 trait a and conditional based on:
```
OpTy::template hasTrait<OpTrait::OneResult>
```
to trigger the `push_back`.

I get the impl of the for loop pushes back in the 1 result case, it still seems like a lot of duplication for a relatively simple switch.

Reply:
You literally asked me to make the non-trait version of this use less templates, but sure.

Reply:
My template-fu is stronger :)